### PR TITLE
feat(minor-coordinator)!: add can unbond utility and connect service registry to coordinator

### PR DESCRIPTION
**Pull request by maancham**
## Description

[AXE-1389]

## Todos

- [ ] Unit tests
- [ ] Manual tests
- [ ] Documentation
- [ ] Connect epics/issues

## Other Notes

We also want to add a multi index map to keep the active/next verifiers for each prover. This will greatly decrease the amount of computation required when coordinator is asked if a verifier is ready to unbond. However, it's not in the scope of this PR (as we need a new execute message and probably an IndexedMap for coordinator and changes to provers confirm and update workerset functions).

[AXE-1389]: https://axelarnetwork.atlassian.net/browse/AXE-1389?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: contracts/coordinator/src/msg.rs
Comment (cjcobb23) [reply contracts/coordinator/src/msg.rs:33]: actually I realized we are not tracking the chains associated with a given worker in this contract. But didn't we want to do that? Wasn't that the whole reason to move to a push model and actually store the active workersets in this contract? Otherwise, we already have the prover address, so why not just query the prover directly each time.

File: contracts/service-registry/src/contract/execute.rs
Comment (cjcobb23) [reply contracts/service-registry/src/contract/execute.rs:177]: I think it does not. I think if a worker calls deregister chain support, the contract will remove them from this map, and then we won't actually check whether the worker is still in the active set. Double check me here.

File: integration-tests/tests/bond_unbond.rs
Comment (cjcobb23) [inline integration-tests/tests/bond_unbond.rs:33]: I think this function just deauthorizes right? If so, can we have a test where the worker is not deauthorized, but simply deregisters chain support? So worker deregisters chain support, and then tries to unbond while still in the active workerset, which should fail.

**Review by cjcobb23 — CHANGES_REQUESTED** [review]

**Comment by cjcobb23** [comment]
Also, I realized we need to handle the case where the worker is not in the active set, but they are in the next worker set that is already in rotation (created at the prover and just awaiting confirmation).

File: contracts/coordinator/src/contract.rs
Comment (cgorenflo) [inline contracts/coordinator/src/contract.rs:55]:
```suggestion
        QueryMsg::ReadyToUnbond {
```
I find "can unbond" a confusing term because it sounds like we're checking if the worker is allowed to unbond. From how the response is used that is not the case: If the result is `false`, the service registry starts the unbonding process for the worker, if it's `true` it completes it. So I advocate for a rename

File: contracts/coordinator/src/msg.rs
Comment (cgorenflo) [reply contracts/coordinator/src/msg.rs:33]: agreed, this check should be very simple: "is this worker ready to unbond?" the coordinator should figure out all the necessary information

File: contracts/service-registry/src/contract/execute.rs
Comment (cgorenflo) [inline contracts/service-registry/src/contract/execute.rs:179]: the service registry shouldn't have to figure out which chains to call

File: contracts/service-registry/src/contract/execute.rs
Comment (cgorenflo) [inline contracts/service-registry/src/contract/execute.rs:188]:
```suggestion
    let ready_to_unbond = deps.querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
        contract_addr: service.coordinator_contract.into(),
        msg: to_binary(&query)?,
    }))?;
```

File: contracts/service-registry/src/contract/execute.rs
Comment (cgorenflo) [inline contracts/service-registry/src/contract/execute.rs:190]: change the argument name inside the `unbond` function to `ready_to_unbond` as well

File: contracts/service-registry/src/state.rs
Comment (cgorenflo) [inline contracts/service-registry/src/state.rs:24]: did the purpose of this field change? What was it and why is it not needed anymore?

File: integration-tests/tests/bond_unbond.rs
Comment (cgorenflo) [inline integration-tests/tests/bond_unbond.rs:1]: do we have a test for unbonding while jailed?

**Comment by cjcobb23** [comment]
I realized there is an edge case that we didn't consider in the design. Consider the case where a workerset update is started at the multisig prover, and a given worker is not in the currently active workerset, but will be in the next workerset (once the update is confirmed). Then, that worker tries to unbond. They will not be in the active workerset yet, so the unbonding is allowed. Then, the new workerset is confirmed at the multisig prover. So now the worker is in the active workerset, but is unbonding. I think we are not handling this case, and we should.

File: contracts/service-registry/src/state.rs
Comment (maancham) [reply contracts/service-registry/src/state.rs:24]: I was told that it was always meant for the coordinator. Since the name was not finalized when creating the instantiation of service registry, it was named as `service_contract`.

File: contracts/coordinator/src/msg.rs
Comment (maancham) [reply contracts/coordinator/src/msg.rs:31]: I changed it to `ReadyToUnbond` based on this comment.

File: contracts/coordinator/src/contract.rs
Comment (maancham) [reply contracts/coordinator/src/contract.rs:55]: I changed it to `ReadyToUnbond` based on this comment.

Comment (maancham) [reply contracts/coordinator/src/contract.rs:55]: Makes sense. Thanks.

File: contracts/coordinator/src/msg.rs
Comment (maancham) [reply contracts/coordinator/src/msg.rs]: Thanks for mentioning it. This is fixed now.

File: contracts/service-registry/src/contract/execute.rs
Comment (maancham) [reply contracts/service-registry/src/contract/execute.rs]: I have migrated the map to coordinator itself. There are not deregister support there and therefore, won't have this issue now.

File: integration-tests/tests/bond_unbond.rs
Comment (maancham) [reply integration-tests/tests/bond_unbond.rs:33]: Added a test for it.

File: integration-tests/tests/bond_unbond.rs
Comment (maancham) [reply integration-tests/tests/bond_unbond.rs:1]: Added!

File: contracts/multisig-prover/src/execute.rs
Comment (maancham) [reply contracts/multisig-prover/src/execute.rs:207]: I changed it and added the message to update_worker_set instead which is the one who is calling the save_next_worker_set.

File: contracts/coordinator/src/query.rs
Comment (cgorenflo) [inline contracts/coordinator/src/query.rs:24]:
```suggestion
pub fn next_worker_set(deps: Deps, chain_name: ChainName) -> StdResult<Option<WorkerSet>> {
```

File: contracts/service-registry/src/state.rs
Comment (cjcobb23) [inline contracts/service-registry/src/state.rs:178]: do we need this data structure at all anymore?

File: contracts/coordinator/src/msg.rs
Comment (cjcobb23) [inline contracts/coordinator/src/msg.rs:27]: Why do we need this method? Can't this info just be inferred from which provers are calling `SetActiveVerifiers` and `SetNextVerifiers`?

File: contracts/coordinator/src/query.rs
Comment (cjcobb23) [inline contracts/coordinator/src/query.rs:58]: you could probably simplify this and write it as just one function:
get the prover address
get the active workerset
get the next workerset
join the sets
check if the worker is in the joined set

File: contracts/coordinator/src/query.rs
Comment (cjcobb23) [inline contracts/coordinator/src/query.rs:25]: I don't think it's safe to unwrap here, the function can return an error

File: contracts/coordinator/src/contract.rs
Comment (cjcobb23) [inline contracts/coordinator/src/contract.rs:65]: I think this needs to be access restricted to the service registry. Right now, anyone can call this with whatever input they like. I think this can be removed though and the info can just be inferred from the above two methods (`SetActiveVerifiers` and `SetNextVerifiers`)

File: contracts/service-registry/src/state.rs
Comment (maancham) [reply contracts/service-registry/src/state.rs:178]: Removed!

File: contracts/coordinator/src/query.rs
Comment (maancham) [reply contracts/coordinator/src/query.rs:58]: I have some unit tests that rely on the individual functions (get active and get next) so we can't remove the functions.

File: contracts/coordinator/src/contract.rs
Comment (maancham) [reply contracts/coordinator/src/contract.rs:65]: Removed

**Review by cjcobb23 — APPROVED** [review]

File: ampd/Cargo.toml
Comment (fish-sammy) [inline ampd/Cargo.toml:38]: Why remove this?

**Review by fish-sammy — CHANGES_REQUESTED** [review]

File: ampd/Cargo.toml
Comment (maancham) [reply ampd/Cargo.toml:38]: There are a ton of conflicts before merge. I talked with Chris. I have all other files on local and will push when we decide on what to do with them. I have actually not removed this file. Marking PR as draft until then...

**Review by cgorenflo — APPROVED** [review]

File: contracts/coordinator/src/query.rs
Comment (maancham) [reply contracts/coordinator/src/query.rs:26]: Yes. But it is an optimization rather that breaking change.

File: contracts/service-registry/src/migrations/v_0_3.rs
Comment (maancham) [reply contracts/service-registry/src/migrations/v_0_3.rs:65]: If I revert the changes in this file, everything fails (all instantiations of service registry). Should I revert all the changes made?

File: contracts/coordinator/src/query.rs
Comment (fish-sammy) [inline contracts/coordinator/src/query.rs:40]: Create a `.includes` method for `VerifierSet`. Also, the comment says that the keys in `.signers` are the signer addresses so why are you loop to check existence here with a `BTreeMap`?

File: contracts/service-registry/src/migrations/v_0_3.rs
Comment (haiyizxx) [reply contracts/service-registry/src/migrations/v_0_3.rs:65]: remove the v0.3 migration so it won't block merge, and have a separate pr for state migration.

File: contracts/coordinator/src/query.rs
Comment (fish-sammy) [inline contracts/coordinator/src/query.rs:40]: Create a `.includes` method for `VerifierSet`. Also, the comment says that the keys in `.signers` are the signer addresses so why are you loop to check existence here with a `BTreeMap`?

File: contracts/coordinator/src/query.rs
Comment (fish-sammy) [inline contracts/coordinator/src/query.rs:35]: Why not use the `active_verifier_set` function above here?

File: contracts/coordinator/src/query.rs
Comment (fish-sammy) [inline contracts/coordinator/src/query.rs:52]: Same here. Looping through the `BTreeMap` defeats the purpose of using a map.

File: contracts/coordinator/src/query.rs
Comment (fish-sammy) [inline contracts/coordinator/src/query.rs:32]:
```suggestion
    let chain_names = PROVER_PER_CHAIN
        .keys(deps.storage, None, None, cosmwasm_std::Order::Ascending)
        .collect::<StdResult<Vec<_>>>()?;

    for chain_name in chain_names {
```

File: contracts/multisig-prover/src/execute.rs
Comment (fish-sammy) [inline contracts/multisig-prover/src/execute.rs:250]: The coordinator contract should have its client mod. Check https://github.com/axelarnetwork/axelar-amplifier/blob/24080b4d91528dfab2d91f8ada5d6e0985262be0/contracts/voting-verifier/src/client.rs#L4. Can do in a separate PR.

**Review by fish-sammy — CHANGES_REQUESTED** [review]

File: contracts/multisig-prover/src/execute.rs
Comment (maancham) [reply contracts/multisig-prover/src/execute.rs:250]: Created a ticket for it.

File: contracts/coordinator/src/query.rs
Comment (maancham) [reply contracts/coordinator/src/query.rs:40]: Added.

File: contracts/service-registry/Cargo.toml
Comment (haiyizxx) [inline contracts/service-registry/Cargo.toml:38]:
```suggestion
coordinator = { workspace = true,  features = ["library"] }
```

File: contracts/coordinator/src/query.rs
Comment (fish-sammy) [reply contracts/coordinator/src/query.rs:35]: What do you mean it's used for tests only? It's used here https://github.com/axelarnetwork/axelar-amplifier/blob/1558bedeb0659a1c8c8b1b88692efa2fd9305f59/contracts/coordinator/src/contract.rs#L75.

File: contracts/multisig/src/verifier_set.rs
Comment (fish-sammy) [inline contracts/multisig/src/verifier_set.rs:65]:
```suggestion
    pub fn includes(&self, signer: &Addr) -> bool {
        self.signers.contains_key(signer.as_str())
    }
```

**Review by fish-sammy — APPROVED** [review]

File: contracts/coordinator/src/query.rs
Comment (fish-sammy) [reply contracts/coordinator/src/query.rs:35]: Then using functions to avoid duplicate code should be the consistency to follow.

**Pull request merged**
